File: java/src/main/java/com/carepet/server/ModelController.java
package com.carepet.server;

import com.carepet.model.*;
import com.datastax.oss.driver.api.core.CqlSession;
import com.datastax.oss.driver.api.core.cql.ResultSet;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.QueryValue;
import io.micronaut.http.exceptions.HttpStatusException;
import io.micronaut.validation.Validated;
import io.reactivex.Observable;
import io.reactivex.Single;

import javax.inject.Inject;
import javax.validation.constraints.NotBlank;
import java.time.*;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Controller("/api")
@Validated
public class ModelController {
    private CqlSession session;
    private Mapper mapper;

    @Inject
    public ModelController(CqlSession session) {
        this.session = session;
        this.mapper = Mapper.builder(session).build();
    }

    private static void groupBy(List<Float> data, List<Measure> measures, int startHour, LocalDate day, LocalDateTime now) {
        // if it's the same day, we can't aggregate current hour
        boolean sameDate = now.getDayOfYear() == day.getDayOfYear();
        int last = now.getHour();

        class Avg {
            double value;
            int total;
        }

        // aggregate data
        Avg[] ag = new Avg[24];

        for (Measure m : measures) {
            int hour = m.getTs().atOffset(ZoneOffset.UTC).getHour();

            if (ag[hour] == null) {
                ag[hour] = new Avg();
            }

            Avg a = ag[hour];
            a.total++;
            a.value += m.getValue();
        }

        // ensure data completeness
        for (int hour = startHour; hour < 24; hour++) {
            if (!sameDate || hour <= last) {
                if (ag[hour] == null) {
                    ag[hour] = new Avg();
                }
            }
        }

        // fill the avg
        for (int hour = startHour; hour < ag.length && ag[hour] != null; hour++) {
            Avg a = ag[hour];
            if (a.total > 0) {
                data.add((float) (a.value / a.total));
            } else {
                data.add(0.0f);
            }
        }
    }

    @Get(uri = "/owner/{id}", produces = MediaType.APPLICATION_JSON)
    public Single<Owner> owner(@NotBlank UUID id) {
        return Single.just(mapper.owner().get(id));
    }

    @Get(uri = "/owner/{id}/pets", produces = MediaType.APPLICATION_JSON)
    public Observable<Pet> pets(@NotBlank UUID id) {
        return Observable.fromIterable(mapper.pet().findByOwner(id));
    }

    @Get(uri = "/pet/{id}/sensors", produces = MediaType.APPLICATION_JSON)
    public Observable<Sensor> sensors(@NotBlank UUID id) {
        return Observable.fromIterable(mapper.sensor().findByPet(id));
    }

    @Get(uri = "/sensor/{id}/values", produces = MediaType.APPLICATION_JSON)
    public Observable<Float> values(@NotBlank UUID id, @NotBlank @QueryValue String from, @NotBlank @QueryValue String to) {
        ResultSet res = mapper.measurement().find(id, Instant.parse(from), Instant.parse(to));
        return Observable.fromIterable(res.map(x -> x.getFloat(0)));
    }

    @Get(uri = "/sensor/{id}/values/day/{day}", produces = MediaType.APPLICATION_JSON)
    public Observable<Float> avg(@NotBlank UUID id, @NotBlank String day) {
        LocalDate date = LocalDate.parse(day);
        if (date.isAfter(LocalDate.now())) {
            throw new HttpStatusException(HttpStatus.BAD_REQUEST, "request into the future");
        }

        ResultSet res = mapper.sensorAvg().find(id, date);
        List<Float> data = res.map(x -> x.getFloat(0)).all();

        if (data.size() != 24) {
            data = new ArrayList<>(data);
            aggregate(id, date, data);
        }

        return Observable.fromIterable(data);
    }

    public void aggregate(UUID id, LocalDate day, List<Float> data) {
        LocalDateTime now = LocalDateTime.now(Clock.systemUTC());

        // can't aggregate data for post today's date
        if (day.getDayOfYear() > now.getDayOfYear()) {
            throw new HttpStatusException(HttpStatus.BAD_REQUEST, "request into the future");
        }

        // we can start from next missing hour. hours = [0, 23]. len = [0, 24]
        int startHour = data.size();
        Instant startDate = day.atStartOfDay().toInstant(ZoneOffset.UTC);
        Instant endDate = day.atTime(23, 59, 59, 999999999).toInstant(ZoneOffset.UTC);

        List<Measure> measures = mapper.measurement().findWithTimestamps(id, startDate, endDate).
                map(row -> new Measure(null, row.getInstant(0), row.getFloat(1))).
                all();

        int prevAvgSize = data.size();
        groupBy(data, measures, startHour, day, now);

        saveAggregate(id, data, prevAvgSize, day, now);
    }

    // saveAggregate saves the result monotonically sequentially to the database
    private void saveAggregate(UUID sensorId, List<Float> data, int prevAvgSize, LocalDate day, LocalDateTime now) {
        // if it's the same day, we can't aggregate current hour
        boolean sameDate = now.getDayOfYear() == day.getDayOfYear();
        int current = now.getHour();

        for (int hour = prevAvgSize; hour < data.size(); hour++) {
            if (sameDate && hour >= current) {
                break;
            }

            mapper.sensorAvg().create(new SensorAvg(sensorId, day, hour, data.get(hour)));
        }
    }


    public void close() {
        session.close();
    }
}

File: csharp/Server/ModelController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using Cassandra;
using CarePet.Model;
using System.Net;
using System.Threading.Tasks;

namespace CarePet.Server
{
    [ApiController]
    [Route("api")]
    public class ModelController : ControllerBase
    {
        private readonly ISession _session;
        private readonly Mapper _mapper;
        private readonly ILogger<ModelController> _logger;

        public ModelController(ISession session, ILogger<ModelController> logger)
        {
            _session = session;
            session.ChangeKeyspace(Config.Keyspace);
            _mapper = new Mapper(session);
            _logger = logger;
        }

        private static void GroupBy(List<float> data, List<Measure> measures, int startHour, DateTime day, DateTime nowUtc)
        {
            bool sameDate = nowUtc.Date == day.Date;
            int lastHour = nowUtc.Hour;

            var ag = new (double value, int total)?[24];

            foreach (var m in measures)
            {
                int hour = m.Ts.UtcDateTime.Hour;

                if (!ag[hour].HasValue)
                    ag[hour] = (0, 0);

                var entry = ag[hour].Value;
                entry.total++;
                entry.value += m.Value;
                ag[hour] = entry;
            }

            for (int hour = startHour; hour < 24; hour++)
            {
                if (!sameDate || hour <= lastHour)
                {
                    if (!ag[hour].HasValue)
                        ag[hour] = (0, 0);
                }
            }

            for (int hour = startHour; hour < ag.Length && ag[hour].HasValue; hour++)
            {
                var entry = ag[hour].Value;
                data.Add(entry.total > 0 ? (float)(entry.value / entry.total) : 0f);
            }
        }

        [HttpGet("owner/{id}")]
        public ActionResult<Owner> Owner(Guid id)
        {
            var owner = _mapper.Owner().Get(id);
            if (owner == null)
                return NotFound();
            return Ok(owner);
        }

        [HttpGet("owner/{id}/pets")]
        public ActionResult<IEnumerable<Pet>> Pets(Guid id)
        {
            return Ok(_mapper.Pet().FindByOwner(id));
        }

        [HttpGet("pet/{id}/sensors")]
        public ActionResult<IEnumerable<Sensor>> Sensors(Guid id)
        {
            return Ok(_mapper.Sensor().FindByPet(id));
        }

        [HttpGet("sensor/{id}/values")]
        public async Task<ActionResult<IEnumerable<float>>> Values(Guid id, [FromQuery] string from, [FromQuery] string to)
        {
            var resultSet = await _mapper.Measurement().FindAsync(id, DateTimeOffset.Parse(from).UtcDateTime, DateTimeOffset.Parse(to).UtcDateTime);
            var values = resultSet.Select(row => row).ToList();
            return Ok(values);
        }

        [HttpGet("sensor/{id}/values/day/{day}")]
        public async Task<ActionResult<IEnumerable<float>>> Avg(Guid id, string day)
        {
            var date = DateTime.Parse(day).Date;
            if (date > DateTime.UtcNow.Date)
            {
                return BadRequest("request into the future");
            }

            var resultSet = await _mapper.SensorAvg().FindAsync(id, date);
            var data = resultSet.Select(row => row).ToList();

            if (data.Count != 24)
            {
                data = new List<float>(data);
                Aggregate(id, date, data).GetAwaiter().GetResult();
            }

            return Ok(data);
        }

        public async System.Threading.Tasks.Task Aggregate(Guid id, DateTime day, List<float> data)
        {
            var nowUtc = DateTime.UtcNow;

            if (day > nowUtc.Date)
            {
                throw new ArgumentException("request into the future");
            }

            int startHour = data.Count;
            var startDate = new DateTimeOffset(day, TimeSpan.Zero);
            var endDate = new DateTimeOffset(day.AddHours(23).AddMinutes(59).AddSeconds(59).AddTicks(9999999), TimeSpan.Zero);

            var rows = await _mapper.Measurement()
                .FindWithTimestampsAsync(id, startDate.UtcDateTime, endDate.UtcDateTime);

            var measures = rows
                .Select(row => new Measure(id, row.Ts, row.Value))
                .ToList();

            int prevSize = data.Count;
            GroupBy(data, measures, startHour, day, nowUtc);
            SaveAggregate(id, data, prevSize, day, nowUtc);
        }

        private void SaveAggregate(Guid sensorId, List<float> data, int prevSize, DateTime day, DateTime nowUtc)
        {
            bool sameDate = nowUtc.Date == day.Date;
            int currentHour = nowUtc.Hour;

            for (int hour = prevSize; hour < data.Count; hour++)
            {
                if (sameDate && hour >= currentHour)
                    break;

                _mapper.SensorAvg().CreateAsync(new SensorAvg(sensorId, day, hour, data[hour]));
            }
        }

        [NonAction]
        public void Close()
        {
            _session.Dispose();
        }
    }
}
